xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guarenteed.
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def handler(self, callable):
```

Comment:
Right, you use it on event handler, it is not a handler itself.  Another take is that functions should be verbs, and this is a noun.  `handle()` would also be a reasonable alternative to `register()`